Core/clim-core/presentation-defs.lisp

```diff
@@ -1610,101 +1610,118 @@ protocol retrieving gestures from a provided string."))
       ;; Now prune out all completions that don't start with the string
       (let ((type (pathname-type actual-pathname)))
-	(when (null type)
-	  ;; If the user didn't supply a file type, don't burden him with all
-	  ;; sorts of version numbers right now.
-	  (let ((new-completions nil))
-	    (dolist (pathname completions)
-	      (cond
-		;; meaning this is actually a directory
-		((and (null (pathname-name pathname))
-		      (null (pathname-type pathname)))
-		 (pushnew (make-pathname :host (pathname-host original-pathname)
-					 :device (pathname-device original-pathname)
-					 :directory (butlast (pathname-directory pathname))
-					 :name (first (last (pathname-directory pathname)))
-					 :type nil)
-			  new-completions))
-		(t
-		 (pushnew (make-pathname :host (pathname-host original-pathname)
-					 :device (pathname-device original-pathname)
-					 :directory (pathname-directory original-pathname)
-					 :name (pathname-name pathname)
-					 :type (pathname-type pathname))
-			  new-completions))))
-	    (setq completions (nreverse new-completions))))
-	(complete-from-possibilities original-string completions '(#\space)
-						    :action action
-						    :name-key #'namestring
-						    :value-key #'identity)))))
+        (when (null type)
+          (flet ((legitimate-logical-pathname (name)
+                   (let ((word (string name)))
+                     (loop for i below (length word)
+                        for ch = (schar word i)
+                        always (and (standard-char-p ch)
+                                    (or (alphanumericp ch) (char= ch #\-)))))))
+            ;; If the user didn't supply a file type, don't burden him with all
+            ;; sorts of version numbers right now.
+            (let ((new-completions nil))
+              (loop for pathname in completions
+                 do (cond
+                      ;; meaning this is actually a directory
+                      ((and (null (pathname-name pathname))
+                            (null (pathname-type pathname)))
+                       (when (and (loop for word in  (butlast (pathname-directory pathname))
+                                     always (legitimate-logical-pathname word))
+                                  (legitimate-logical-pathname (first (last (pathname-directory pathname)))))
+                         (pushnew (if (typep original-pathname 'logical-pathname)
+                                      (make-pathname :host (pathname-host original-pathname)
+                                                     :device (pathname-device original-pathname)
+                                                     :directory (first (last (pathname-directory pathname)))
+                                                     :name nil
+                                                     :type nil)
+                                      (make-pathname :host (pathname-host original-pathname)
```

Comment:
instead of duplicating make-pathname form introduce local variables name and directory and setf them, then do a single pushnew